src/main/scala/io/github/pauljamescleary/petstore/infrastructure/endpoint/UserEndpoints.scala

```diff
@@ -1,28 +1,59 @@
-package io.github.pauljamescleary.petstore.infrastructure.endpoint
+package io.github.pauljamescleary.petstore
+package infrastructure.endpoint
 
+import cats.data.EitherT
 import cats.effect.Effect
 import cats.implicits._
 import io.circe.generic.auto._
 import io.circe.syntax._
+import io.github.pauljamescleary.petstore.domain.authentication.CryptService
 import org.http4s.circe._
 import org.http4s.dsl.Http4sDsl
-import org.http4s.{EntityDecoder, HttpService}
+import org.http4s.{EntityDecoder, HttpService, Response}
 
 import scala.language.higherKinds
-import io.github.pauljamescleary.petstore.domain.{UserAlreadyExistsError, UserNotFoundError}
-import io.github.pauljamescleary.petstore.domain.users.{User, UserService}
+import domain._
+import domain.users._
+import domain.authentication._
+import tsec.authentication._
+import tsec.passwordhashers.PasswordHash
 
-class UserEndpoints[F[_]: Effect] extends Http4sDsl[F] {
+class UserEndpoints[F[_]: Effect, A, K] extends Http4sDsl[F] {
+  type AuthService = AuthenticatorService[F, Long, User, K]
+  type CS = CryptService[F, A]
 
   import Pagination._
   /* Jsonization of our User type */
-  implicit val userDecoder: EntityDecoder[F, User] = jsonOf[F, User]
+  implicit val userDecoder: EntityDecoder[F, User] = jsonOf
+  implicit val loginReqDecoder: EntityDecoder[F, LoginRequest] = jsonOf
 
-  private def signupEndpoint(userService: UserService[F]): HttpService[F] =
+  implicit val signupReqDecoder: EntityDecoder[F, SignupRequest] = jsonOf
+
+  private def loginEndpoint(userService: UserService[F], crypt: CS, Auth: AuthService): HttpService[F] =
+    HttpService[F] {
+      case req @ POST -> Root / "login" =>
+        val action: EitherT[F, UserAuthenticationFailedError, Response[F]] = for {
+          login <- EitherT.liftF(req.as[LoginRequest])
+          name = login.userName
+          user <- userService.getUserByName(name).leftMap(_ => UserAuthenticationFailedError(name))
+          valid <- EitherT.liftF(crypt.check(login.password, PasswordHash[A](user.hash)))
+          resp <- EitherT.liftF(Ok())
+        } yield resp
+
+        // Handle error cases:
+        action.value.flatMap{
+          case Right(resp) => resp.pure[F]
+          case Left(UserAuthenticationFailedError(name)) => Conflict(s"Authentication failed for user $name")
```

Comment:
Is `Conflict` the correct response?

Reply:
No. I think 401 is right, but I couldn't figure out how to instantiate it, the `apply` on `Unauthorized` is confusing, since a 401 requires some WWW-Authenticate header. Maybe just `400`/`BadRequest`